### PR TITLE
Merge Backward and Optimize

### LINKED ISSUE #48: merge backward and optimizer
```c++
void BackwardAndUpdate(VariableHandle loss, Optimizer& opt) {
	get_global_tape().Backward(loss);
	opt.Step();
}

class Optimizer {
public:
	virtual void Step() = 0;
	virtual void Update(ParameterHandle param) = 0;
};

class SGD : public Optimizer {
public:
	virtual void Step() override {
		for (auto& w : AllParameters()) {
			Update(w);
		}
	}
};

class Adam : public Optimizer {
public:
	virtual void Step() override {
		UpdateBeta();
		for (auto& w : AllParameters()) {
			Update(w);
		}
	}
};
```

### PR DESCRIPTION
**Pull request by kexinzhao**
fix #48 

Also modified adam optimizer implementation to save host-device memory copy.

### CODE REVIEW

File: src/tape.h [hunk 1]
```diff
@@ -122,6 +134,7 @@ class Tape {
              const framework::AttributeMap &attrs);
   void Forward();
   void Backward(VariableHandle target);
+  void BackwardAndUpdate(VariableHandle target, Optimizer *optimizer);
```
Comment (tonyyang-svail) [inline src/tape.h]: We should keep `tape.h` as simple as possible. Tape only manages the backward, so don't add optimizer's logic at `tape.h`.

A simple function might be sufficient.
```c++
void BackwardAndUpdate(VariableHandle loss, Optimizer& opt) {
	get_global_tape().Backward(loss);
	opt.Step();
}
```

File: src/tape.h [hunk 2]
```diff
@@ -31,8 +31,20 @@ class Variable;
 using VariableHandle = std::shared_ptr<Variable>;
 using VariableHandleMap = std::map<std::string, std::vector<VariableHandle>>;
 
+class Parameter;
+using ParameterHandle = std::shared_ptr<Parameter>;
+
 std::ostream &operator<<(std::ostream &, const Variable &);
 
+class Optimizer {
```
Comment (tonyyang-svail) [inline src/tape.h]: We should keep `tape.h` as simple as possible. Tape only manages the backward, so don't add optimizer's logic at `tape.h`.

File: src/tape.h [hunk 3]
```diff
@@ -31,8 +31,20 @@ class Variable;
 using VariableHandle = std::shared_ptr<Variable>;
 using VariableHandleMap = std::map<std::string, std::vector<VariableHandle>>;
 
+class Parameter;
```
Comment (tonyyang-svail) [inline src/tape.h]: We should keep `tape.h` as simple as possible. Tape only manages the backward, it doesn't care if the variable is a `Parameter`

**Review by tonyyang-svail — CHANGES_REQUESTED** [review]
Optimizer may depend on Tape, Tape should not depend on Optimizer.

File: src/tape.h (hunk 1 above)
Comment (kexinzhao) [reply src/tape.h]: Done. Made a small change, as this optimizer might change state inside this function. So use pointer instead.

File: src/tape.h (hunk 2 above)
Comment (kexinzhao) [reply src/tape.h]: Done.

File: src/tape.h (hunk 3 above)
Comment (kexinzhao) [reply src/tape.h]: Done.

**Review by tonyyang-svail — APPROVED** [review]
Excellent. Maybe add a line in cmake `cc_library(tape_optimizer SRCS optimizer.cc DEPS tape tape_parameter)`. So that we will sure the .h file can be self-compiled. :)

**Comment by kexinzhao** [comment]
Thanks! Will change cmake in the next pr.

**Pull request merged**